### PR TITLE
feat: add foundation of error handling logic

### PR DESCRIPTION
**Pull request by 0xyaco**
# 🤖 Linear

Closes GRT-XXX

## Description

### CODE REVIEW

File: packages/automated-dispute/src/services/eboActor.ts [hunk 1]
```diff
@@ -265,15 +292,32 @@ export class EboActor {
...
+                    blockNumber,
+                    registry: this.registry,
+                });
+
+                await ErrorHandler.handle(customError, {
+                    terminateActor: () => {
+                        // TODO: correct termination logic
+                        this.registry.removeRequest(this.actorRequest.id);
+                    },
+                });
```
Comment (0xyaco) [inline packages/automated-dispute/src/services/eboActor.ts:309]: Let's move this inside the function to settle a single dispute `settleDispute`.

File: packages/automated-dispute/src/services/eboActor.ts [hunk 2]
```diff
@@ -149,15 +150,41 @@ export class EboActor {
...
+                        // TODO: Add all errors appropriately -- this is boilerplate
+                        const customError = ErrorFactory.createError(err.name)
+                            .setContext({ event, registry: this.registry })
+                            .on("BondEscalationModule_ShouldBeEscalated", async (context) => {
+                                await this.protocolProvider.escalateDispute(
+                                    context.request.prophetData,
+                                    context.response.prophetData,
+                                    context.dispute.prophetData,
+                                );
+                            });
```
Comment (0xyaco) [inline packages/automated-dispute/src/services/eboActor.ts:163]: What do you think about not using `setContext` nor `on` at this stage? I feel like they should solely be used inside the `onXXX` handlers as inside those functions we have the whole RPC context that we might need. 

I like that you added the `event` to the context and the `blockNumber` in the other cases.

File: packages/automated-dispute/src/services/eboActor.ts [hunk 3]
```diff
@@ -149,15 +150,41 @@ export class EboActor {
...
+                                }
+                            },
+                            terminateActor: () => {
+                                // TODO: correct termination logic
+                                this.registry.removeRequest(this.actorRequest.id);
+                            },
+                        });
+
+                        if (customError.strategy.scenario === ErrorScenario.Unrecoverable) {
+                            updateStateCommand.undo();
```
Comment (0xyaco) [inline packages/automated-dispute/src/services/eboActor.ts:181]: This `undo()` should go in the `reenqueEvent` handler, not here

File: packages/automated-dispute/src/services/eboActor.ts [hunk 4]
```diff
@@ -149,15 +150,41 @@ export class EboActor {
                         await this.onLastEvent(event);
                     }
                 } catch (err) {
-                    this.logger.error(`Error processing event ${event.name}: ${err}`);
```
Comment (0xyaco) [inline packages/automated-dispute/src/services/eboActor.ts:152]: Let's keep this log as a general log for all errors

File: packages/automated-dispute/src/services/eboActor.ts [hunk 5]
```diff
@@ -282,9 +326,30 @@ export class EboActor {
...
+                if (err instanceof ContractFunctionRevertedError) {
+                    const customError = ErrorFactory.createError(err.name).setContext({
+                        request,
+                        acceptedResponse,
+                        registry: this.registry,
+                    });
+
+                    await ErrorHandler.handle(customError, {
+                        terminateActor: () => {
+                            // TODO: correct termination logic
```
Comment (0xyaco) [inline packages/automated-dispute/src/services/eboActor.ts:344]: Remember that if the actor throws an error outside its scope, the `EboProcessor` will catch it and terminate the actor. The actor is supposed to handle all errors except the critical ones, in which case it throws. No registry manipulation needed

File: packages/automated-dispute/src/types/errorTypes.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+export enum ErrorScenario {
+    Skippable = "Skippable",
+    Retryable = "Retryable",
+    Unrecoverable = "Unrecoverable",
+}
```
Comment (0xyaco) [inline packages/automated-dispute/src/types/errorTypes.ts:5]: I think that we can drop this `ErrorScenario` enum. Those values can be derived by the `shouldXXX` properties defined in each error's strategy. I couldn't find a place where these are necessary.

**Review by 0xyaco — COMMENTED** [review]
Looking good! Let's do a small PoC for the following scenarios based on the error specs:
* Dispute Response → ValidatorLib_InvalidResponseBody
* Dispute Response → BondEscalationAccounting_InsufficientFunds
* Dispute Response → BondEscalationAccounting_AlreadySettled
* Pledge For or Against Dispute → BondEscalationModule_InvalidDispute
* Settle Dispute → AccountingExtension_InsufficientFunds
* Settle Dispute → Oracle_InvalidDisputeId
* Finalize Request → Oracle_AlreadyFinalized
* Finalize Request → EBORequestModule_InvalidRequester

**Review by 0xyaco — COMMENTED** [review]
I'd say as a general rule, given that we have two scenarios:
* Errors while reacting to events
* Errors while reacting to time-based windows

For the first one, we should **build** the error inside the `onXXX` function, throw it and handle it inside the `processEvents` `try-catch` block. This is because event related actions (retry or consume) needs to be handled inside the `processEvents` function.

For the second scenario, we should **build and handle** the error inside `settleDispute` and a potential `finalizeRequest` function (probably worth extracting all the related code into its own function as now it's all splattered inside `onLastBlockUpdated`)

**Review by 0xyaco — COMMENTED** [review]
Looking super sleek!

---

**Question 1**
For _consuming events_ (ie popping them from the event queue and not re-enqueueing them even when the RPC method reverts) we could just use `!shouldReenqueue`. Let's remove `shouldConsume` from the error type and just use `shouldReenqueue` as the only thing we want this property for is to know if the event should be reenqueued when an error has been thrown.

TL;DR `EboActor` should leverage `shouldReenqueue` error property only when handling errors thrown when calling `onLastEvent`.

---

**Question 2**
Reacting to those two events (`DisputeEscalated` and `RequestFinalized`) are only for informative purposes (ie updating internal state/notifying). There's a high chance of no RPC method being executed while reacting to them.
